### ISSUE TITLE
Add support and documentation to extend core classes (eg. Walker class) 

### ISSUE THREAD

**Issue by benoitchantre** [wprig/wprig#560]
## Issue Overview

WP Rig components are loaded by an autoloader. If you need to extend a core class, this is not supported. Extending core classes should be supported and documented.

Exemple of use case : walker classes.

Cc. @felixarntz, @ataylorme

**Comment by Horlaes** [comment]
I have a problem using the walker class at the moment, any work around?

**Comment by mor10** [comment]
This needs some serious work

**Comment by justlevine** [comment]
Im super not familiar with Nav Walkers, but I was able to create a widget class really simply.

Just opened it with `class Component extends \WP_Widget implements Component_Interface`, and added `__construct()` __in adition__ to initialize to call `parent::_construct( 'widget-slug', 'My Widget Title', $widget_ops );

Will something similar not work for a walker?

(PS: completely agree regarding the lack of documentation. This theme does things quite differently than I'm used to, the tutorial videos dont replace the need for actual written and refrensable documentation, and theyre all V1 anyway).

**Comment by justlevine** [comment]
@ataylorme its currently pinned in chrome for easy reference :-P 

My PHP knowledge comes (sadly) mainly from Wordpress, so tbh getting comfortable with class structures and autoloading is my biggest hurdle, and where I think example code snippets could help a lot with the barrier to entry.

Figuring out that I could have both `extends \WP_Widget` _and_ `implements Component_Interface` without mucking up WP_Rigs automagical process, was a few hours of googling and trial and error. Another hour to figure out that I needed to create a `__construct` because `initialize()` wouldnt handle it correctly. And the lack of code examples and my own knowledge means that my [Imposter Syndrome](https://en.wikipedia.org/wiki/Impostor_syndrome) is acting up at full force.

(Re the original topic:  I've rarely had to dabble in nav_walkers, but even if its as 'simple' as adding an `extends /WP_CoreClass` to someone versed in 'real' PHP, combining it with `implements Component_Interface` isnt necessarily obvious to those of us who have never `implement`ed anything before using WP_Rig. Maybe @benoitchantre is in the same boat I am, or maybe I just got lucky that it works with widgets. I honestly dont know, and I think thats part of the issue).

In addition to extending Core classes, other things Ive struggled with are:
- instantiating classes. E.g. Today I tried to add a custom woocommerce endpoint, and while most `[common code examples](https://gist.github.com/claudiosanches/a79f4e3992ae96cb821d3b357834a005)`  will add a `new MyEndpointName()` outside of the class. it took me a while to realize thats the same as adding my `Component()` to `Theme.php`. 
- Converting other WP code over. I gave up on trying to rewrite a Template Wrapper (from [sage 8.x](http://scribu.net/wordpress/theme-wrappers.html), and just added an ugly `require_once` at the the top of `functions.php`. Can I bring functions from outside the class inside? When is `__construct` not necessary and you can solely rely on `initialize()`? I still dont know.
- Autoloading in general.  E.g. Im bundling ACF pro in my theme, composer alone wasnt enough to get it working, so I made a new ACF\Component with `require_once path/to/acf.php` in the `initialize()`. Is this the best way? No idea.
Similarly, subclasses. Ive got a class Post_Types which has some generic template tags, and then subclasses for Post_Types\Events and Post_Types\Listings. Should I be requiring the subclasses in Post_Types, or adding all three to `Theme.php`? 

**TL;DR** (and apologies - I wrote _a lot_): a lot of what Im struggling with might seem fairly basic (like adding `extends \WP_Widget` to get a core class to work), but because of the cognitive load that comes with doing things a new and for many WP Developers ( <= PHP Developers),  Im left wondering both what I _can_ do and whats the prescribed way to do it. IMHO the best way to address it is by offering a few non-standard examples.

And again, if extending walkers are different than extending widgets, than maybe I'm the odd one out and most of this is moot ::shrug::

**Comment by justlevine** [comment]
>I'm not sure WP Rig documentation on its own can get you over this hurdle but one of the goals of the project is to help with the transition.

Agreed wholeheartedly, and even if it could, it's probably shouldn't.

Which is why instead of an explanation, I think the best way forward would be to provide a few example classes for reference. I don't need an explanation on how `extends` and `implements` work, if there's an example widget class that shows how someone would extend a core WP class IRL (or require a php file that isn't a Component I wrote, or uses both construct and initialize).

Thanks for the book recommendation :)

**Comment by Horlaes** [comment]
@justlevine I agree with you 100%, When I started with wp-rig, I struggled for weeks upon weeks getting it up and running, don't even get me started with the **PHP architecture in Wp Rig**, everything here is isolated, I was able to get it working using trial and error, most times no luck, sometimes, I might be lucky, until I mastered my way around it.

Most people would probably give up after several attempts at getting a particular function to work, it's not just for newbies, I was about to give up too, I can't Google Search a specific problem, no much of example in the Wiki page, I remember trying to run a loop In my content page, after days of trying, I noticed if you wanna use any Wordpress core file, you have to import them in your current template, as well as other functions you plan on using.  

Once you understand your way around it, everything becomes easy, another thing I like is the way things are organized, lots of features, minimize, minify on the fly, etc.

I suggest Wp-Rig allow developers to submit snippet on the wp-rig website itself, where users would be able to upvote and downvote base on the quality of the code, that way, users would find it easy to copy snippet or copy steps of achieving a certain feature or functions.

**Comment by felixarntz** [comment]
Thanks for all this feedback. I agree there's a lot we can do to improve our documentation, and that's one of our primary focuses these days.

I think partly it would be great to gather a solid set of third-party resources that document how some of the practices WP Rig's PHP code uses work in general. This would then provide a stepping stone to following the [PHP Architecture documentation](https://github.com/wprig/docs/blob/master/documentation/php.md).

The other part where we can improve is, as suggested multiple times here, to have more example code. @ataylorme had the idea of including some components with common functionality in the `/optional/inc` directory. That way WP Rig would already include some of this sample code to learn from, and you could just move it into the actual `/inc` directory to make them work.

I'm also exploring how the loading process of components and dependencies between them can be simplified overall. It would be much nicer if you could just drop a component folder and class in `/inc` and it would work, rather than having to also manually change the main `Theme` class.

**Comment by felixarntz** [comment]
One more specific comment, related to one of your issues @justlevine:

> Figuring out that I could have both `extends \WP_Widget` and `implements Component_Interface` without mucking up WP_Rigs automagical process, was a few hours of googling and trial and error. 

I believe the idea to have a widget that is also a component proves that we need more examples - especially for one key misunderstanding: Every folder in `/inc` should have only one `Component` class, but you can have as many classes (and thus files) in a single folder as you like. The `Component` class is the entry point, and it could load as many classes as you require from there. It doesn't need to all be in the one `Component` class.

For registering widgets, I'd recommend having a component that is only responsible for registering the widgets - it shouldn't need to be the widget itself. You could add a `WP_Rig\WP_Rig\Widgets\Component` class, whose only purpose it would be to register widgets - in other words, it would act as some kind of widgets manager (because even if you only create one widget initially, you might add more later). The class would only contain a single hook added to `widgets_init`, and in that hook method you would register your widgets. The individual widgets would still be separate classes in the same folder (e.g. `WP_Rig\WP_Rig\Widgets\My_Widget`), and these you would implement in the exact same way you would do in a traditional theme (e.g. `My_Widget extends \WP_Widget`). To register that widget in your `WP_Rig\WP_Rig\Widgets\Component` class, you would need to use `register_widget( My_Widget::class )` (the `::class` syntax resolves a local class name to the full class name, see https://www.php.net/manual/en/language.oop5.basic.php#language.oop5.basic.class.class).

**Comment by prikkprikkprikk** [comment]
> Im bundling ACF pro in my theme, composer alone wasnt enough to get it working, so I made a new ACF\Component with require_once path/to/acf.php in the initialize(). Is this the best way? No idea, but that approach doesnt sound like 'autoloading'.

@justlevine Did you get ACF to work that way? Care to share some code? Given the prevalence of ACF, it would be good to have an article about this on the web somewhere.

**Comment by noeldevelops** [comment]
> > Im bundling ACF pro in my theme, composer alone wasnt enough to get it working, so I made a new ACF\Component with require_once path/to/acf.php in the initialize(). Is this the best way? No idea, but that approach doesnt sound like 'autoloading'.
> 
> @justlevine Did you get ACF to work that way? Care to share some code? Given the prevalence of ACF, it would be good to have an article about this on the web somewhere.

I also want to know. I've been searching for how to get ACF to work with this theme for hours. @justlevine are you still using WP Rig?

**Comment by justlevine** [comment]
Sorry I didn't see this back in December. 

WPRig's autoload doesn't use composer's autoload in production, so yeah if you're bundling, then the way to do it is to create Plugins\ACF\Component() with the the `initialize()` including `require_once` pointed to acf.php

Personally, I leave acf as a recommended plugin, and then wrap the ` new ACF/Component()` in a `class_exists()` and then just have the Component class hold the actions, filters and plugin settings.

Since plugins load up before themes, I find this limits unexpected behavior when trying to use certain acf actions/filters.

**Comment by justlevine** [comment]
This is a the barebones starting point I usually use as a starting point for ACF: https://gist.github.com/justlevine/405c9d5eceee64f5260c20597df78412

All it really does is create a site-options page, stores acf site options in a transient and the page options to a variable (to avoid all those horribly slow and repetitive `get_fields()` calls)...

**Comment by noeldevelops** [comment]
@justlevine thank you so much!

**Comment by fsdevriver** [comment]
Are there any updates about extending core classes?
